python-as-wrapper.yaml

```diff
@@ -0,0 +1,176 @@
+package:
+  name: python-as-wrapper
+  version: 0.1.0
+  epoch: 0
+  description: Shell script wrapper to exec python3.XX
+  copyright:
+    - license: Apache-2.0
+
+# In a "normal" installation of python, 'python3' is a symlink
+# to 'pythonX.Y' where X.Y is Major.Minor (3.13).  In that scenario,
+# if caller invokes 'python3', then when python reads its sys.executable,
+# the executable is 'python3'.
+#
+# That value of sys.executable is "sticky" and finds its way into shbangs that
+# are installed with 'pip install' or other places.
+#
+# In the vast majority of places, we do not want shebang with
+# /usr/bin/python3 because at runtime:
+#  * python3 might be a symlink to another python version
+#  * python3 may not exist at all - with only python3.XX installed.
+#
+# Using the wrapper that this package provides if you do
+# 'python3 -m pip install pkg', then 'pkg's provided scripts will have
+# a shbang of /usr/bin/python3.XX.
+data:
+  - name: py-versions
+    items:
+      3.10: '310'
+      3.11: '311'
+      3.12: '312'
+      3.13: '313'
+  - name: python-as-names
+    items:
+      python: PYTHON_AS
+      python3: PYTHON3_AS
+
+environment:
+  contents:
+    packages:
+      - busybox
+
+pipeline:
+  - runs: |
+      cat > write-strict <<"EOF"
+      #!/bin/sh -e
+      input="$1"
+      outf="$2"
+      mkdir -p "${outf%/*}"
+      printf '#!/bin/sh\nexec "%s" "$@"\n' "$input" > "$outf"
+      chmod 755 "$outf"
+      EOF
+      chmod 755 write-strict
+
+      cat > wrapper-env.tmpl <<"EOF"
+      #!/bin/sh
+      fail() { echo "FATAL:" "$0" "$@"; exit 9; }
+      [ -n "${ENV_NAME}" ] || fail "invoked without ENV_NAME set"
+      [ "$ENV_NAME" -ef "$0" ] && fail "invoked with ENV_NAME = $0"
+      [ "${ENV_NAME}" != "${0##*/}" ] || fail "invoked with ENV_NAME = basename($0)"
+      command -v "$ENV_NAME" >/dev/null ||
+        fail "invoked with ENV_NAME=$ENV_NAME, but $ENV_NAME not found in PATH"
+      exec "${ENV_NAME}" "$@"
+      EOF
+
+      cat > write-env <<"EOF"
+      #!/bin/sh -e
+      input="$1"
+      outf="$2"
+      d=$(dirname "$0")
+      tmpl="$d/wrapper-env.tmpl"
+      [ -f "$tmpl" ] || { echo "no 'wrapper-env.tmpl' in dir '$d/'"; exit 1; }
+      mkdir -p "${outf%/*}"
+      sed "s,ENV_NAME,$input,g" "$tmpl" > "$outf"
+      chmod 755 "$outf"
+      EOF
+      chmod 755 write-env
+
+subpackages:
+  - range: py-versions
+    name: python3-as-${{range.key}}
+    description: provides /usr/bin/python3 that execs python${{range.key}}
+    dependencies:
+      runtime:
+        - python-${{range.key}}-base
+    pipeline:
+      - runs: |
+          ./write-strict "python${{range.key}}" ${{targets.contextdir}}/usr/bin/python3
+    test:
+      pipeline:
+        - name: validate wrapper 'python3' for ${{range.key}}
```

Comment:
to avoid code duplication, maybe we could put this in a `python-as-wrapper-tests` package and ship it?

Reply:
I did consider that. I can do that if you'd like.

Reply:
done.